certora/applyMunging.patch

```diff
@@ -0,0 +1,132 @@
+diff -ruN DelegationToken.sol DelegationToken.sol
+--- DelegationToken.sol	2024-11-06 18:57:11.000000000 +0100
++++ DelegationToken.sol	2024-11-26 11:36:31.000000000 +0100
+@@ -4,14 +4,14 @@
+ import {IDelegation, Signature, Delegation} from "./interfaces/IDelegation.sol";
+ 
+ import {ERC20PermitUpgradeable} from
+-    "../lib/openzeppelin-contracts-upgradeable/contracts/token/ERC20/extensions/ERC20PermitUpgradeable.sol";
++    "../../lib/openzeppelin-contracts-upgradeable/contracts/token/ERC20/extensions/ERC20PermitUpgradeable.sol";
+ import {ECDSA} from
+-    "../lib/openzeppelin-contracts-upgradeable/lib/openzeppelin-contracts/contracts/utils/cryptography/ECDSA.sol";
++    "../../lib/openzeppelin-contracts-upgradeable/lib/openzeppelin-contracts/contracts/utils/cryptography/ECDSA.sol";
+ import {Ownable2StepUpgradeable} from
+-    "../lib/openzeppelin-contracts-upgradeable/contracts/access/Ownable2StepUpgradeable.sol";
+-import {UUPSUpgradeable} from "../lib/openzeppelin-contracts-upgradeable/contracts/proxy/utils/UUPSUpgradeable.sol";
++    "../../lib/openzeppelin-contracts-upgradeable/contracts/access/Ownable2StepUpgradeable.sol";
++import {UUPSUpgradeable} from "../../lib/openzeppelin-contracts-upgradeable/contracts/proxy/utils/UUPSUpgradeable.sol";
+ import {ERC1967Utils} from
+-    "../lib/openzeppelin-contracts-upgradeable/lib/openzeppelin-contracts/contracts/proxy/ERC1967/ERC1967Utils.sol";
++    "../../lib/openzeppelin-contracts-upgradeable/lib/openzeppelin-contracts/contracts/proxy/ERC1967/ERC1967Utils.sol";
+ 
+ /// @title DelegationToken
+ /// @author Morpho Association
+@@ -44,6 +44,9 @@
+         mapping(address => uint256) _delegationNonce;
+     }
+ 
++    // A ghost variable to track the theoretical voting power of address zero.
++    uint256 _zeroVirtualVotingPower = 0;
++
+     /* ERRORS */
+ 
+     /// @notice The signature used has expired.
+@@ -102,6 +105,19 @@
+         _delegate(delegator, newDelegatee);
+     }
+ 
++    function _delegatorFromSig(Delegation calldata delegation, Signature calldata signature)
++        external
++        returns (address)
++    {
++        address delegator = ECDSA.recover(
++            _hashTypedDataV4(keccak256(abi.encode(DELEGATION_TYPEHASH, delegation))),
++            signature.v,
++            signature.r,
++            signature.s
++        );
++        return delegator;
++    }
```

Comment:
Let's add this to a Harness or a Util solidity file. I think we want to keep the munging as minimal as possible, because it's difficult to see changes there

Reply:
Funny that you should mention this because I've tried that already and it seams that it's not as simple as I thought. I initially implemented it in a `Utils` helper contract that inherited from the `DelegationToken` to no avail (ikd if ecdsa is that sensitive to context), so I figured that this was the simplest solution that would go through. Maybe, I didn't implement it well tough.

Reply:
You can see this [run](https://prover.certora.com/output/129828/d7304a0aefd84a3c9836e30a8849187f/?anonymousKey=b5d8b1fc72f65bf64a07c223bff489499107da28).

Reply:
Yes, it's not completely trivial, because domain separator depends on the contract address (and the contract name & symbol). So the easiest solution would be to have a Harness contract I think, which would ensure that it uses the same address and the same name & symbol

Reply:
Implemented here https://github.com/morpho-org/morpho-token-upgradeable/pull/92/commits/a6bf5a4dc4b0d9796d71c078a666166d20b77ebd. Although, I've implemented the suggestion to try it out, I suggest we roll back to the previous commit. This new commit introduces two new files (the harnesses) and duplicate the code. This means that we now need to maintain two harnesses instead of a single patch. That is, unless we find a way to avoid code duplication, then the only inconvenient would be to add more files to the setup.

Reply:
I agree that it's not ideal, but it does not bother me too much even with the code duplication. I think we can avoid code duplication though by creating a contract on top of `DelegationToken` with the added function, and inheriting from it each time in `MorphoTokenOptimismHarness` and `MorphoTokenEthereumHarness`.

Another solution would be to change `delegateWithSig` to return the `delegator`. It's still munging, but it's lighter I think, although it may be less clear that it's not changing how the code works

Reply:
With latest commit [2ad95dd](https://github.com/morpho-org/morpho-token-upgradeable/pull/92/commits/2ad95ddb75cf2bcb42808b51c4ab677e0b99a474) it starts to be pretty clean and not so much invasive. After some thought, adding another contract to the scene is not that weird, and should be allowed and safe to do